core/validation/validation.go

```diff
@@ -14,28 +14,79 @@ import (
 	"fmt"
 	"github.com/spiffe/go-spiffe/v2/workloadapi"
 	"github.com/vmware-tanzu/secrets-manager/core/env"
+	"regexp"
 	"strings"
 )
 
+const spiffeRegexPrefixStart = "^"
+
 // IsSentinel returns true if the given SPIFFEID is a Sentinel ID.
 // It does this by checking if the SPIFFEID has the SpiffeIdPrefixForSentinel
 // as its prefix.
 func IsSentinel(spiffeid string) bool {
-	return strings.HasPrefix(spiffeid, env.SpiffeIdPrefixForSentinel())
+	prefix := env.SpiffeIdPrefixForSentinel()
+
+	if !strings.HasPrefix(
+		spiffeid,
+		"spiffe://"+env.SpiffeTrustDomain()+"/") {
+		return false
+	}
+
+	if strings.HasPrefix(prefix, spiffeRegexPrefixStart) {
+		re, err := regexp.Compile(prefix)
+		if err != nil {
+			return false
+		}
+		return re.MatchString(spiffeid)
+	}
+
+	return strings.HasPrefix(spiffeid, prefix)
 }
 
 // IsSafe returns true if the given SPIFFEID is a Safe ID.
 // It does this by checking if the SPIFFEID has the SpiffeIdPrefixForSafe
 // as its prefix.
 func IsSafe(spiffeid string) bool {
-	return strings.HasPrefix(spiffeid, env.SpiffeIdPrefixForSafe())
+	prefix := env.SpiffeIdPrefixForSafe()
+
+	if !strings.HasPrefix(
+		spiffeid,
+		"spiffe://"+env.SpiffeTrustDomain()+"/") {
+		return false
+	}
+
+	if strings.HasPrefix(prefix, spiffeRegexPrefixStart) {
+		re, err := regexp.Compile(prefix)
+		if err != nil {
+			return false
+		}
+		return re.MatchString(spiffeid)
+	}
+
+	return strings.HasPrefix(spiffeid, prefix)
 }
 
 // IsWorkload returns true if the given SPIFFEID is a WorkloadIds ID.
 // It does this by checking if the SPIFFEID has the SpiffeIdPrefixForWorkload
 // as its prefix.
 func IsWorkload(spiffeid string) bool {
-	return strings.HasPrefix(spiffeid, env.SpiffeIdPrefixForWorkload())
+	prefix := env.SpiffeIdPrefixForWorkload()
```

Comment:
This should be validated both at the workload, and also at VSecM safe.

A malicious workload can easily fake or bypass configuration-based validations that depend on an environment variable or a mounted config file.

The authentication/validation should be mutual, and we should guard the security of core VSecM components with tight RBAC.

Or an alternative is to use a configuration file that can be accessed through a read-only API. But if the workload is a third-party workload that can opt-out from such validation, it is of little use. -- and the env vars that vsecm components use (such as safe and sentinel, etc) cannot be tampered with unless changing the source code of those components and recompiling them (or having a malicious cluster admin misconfiguring them, which then all bets will be off anyway.)